File: python/palindrome-products/palindrome_products.py
def smallest(min_factor, max_factor):
    return calculate(min_factor, max_factor, False)

def largest(min_factor, max_factor):
    return calculate(min_factor, max_factor, True)

def calculate(low, high, rev):
    if low > high:
        raise ValueError("min must be <= max")

    multiples = range(low**2, high**2+1)
    if rev:
        multiples = reversed(multiples)

    for x in multiples:
        if str(x) == str(x)[::-1]:
            factors = [(y, x/y) for y in range(low, high+1) if x%y==0 and low <= x/y <= high]
            if factors:
                return x, factors

    return None, []

File: javascript/palindrome-products/palindrome-products.js
export class Palindromes {
    static generate({ minFactor, maxFactor }) {
        if (minFactor > maxFactor) {
            throw new Error('min must be <= max');
        }

        const result = {
            smallest: new Result(),
            largest: new Result(),
        };

        const minProduct = minFactor ** 2;
        const maxProduct = maxFactor ** 2;
        const diff = maxProduct - minProduct;

        for (let i = 0; i < diff; i++) {
            const smallProduct = minProduct + i;
            const largeProduct = maxProduct - i;

            if (!result.smallest.done()){
                result.smallest = getPalindrome(minFactor, maxFactor, smallProduct);
            }

            if (!result.largest.done()) {
                result.largest = getPalindrome(minFactor, maxFactor, largeProduct);
            }

            if (result.smallest.done() && result.largest.done()) {
                break;
            }
        }

        return result;
    }
}

class Result {
    constructor(value=null, factors=[]) {
        this.value = value;
        this.factors = factors;
    }

    done() {
        return this.factors.length > 0;
    }
}

function getPalindrome(min, max, prod) {
    if (!isPalindrome(prod)) {
        return new Result();
    }

    const factors = [];

    for (let i = min; i <= max; i++) {
        const div = prod / i;
        if (prod % i === 0 && div <= max && div >= i) {
            factors.push([i, div]);
        }
    }

    return new Result(prod, factors);
}

function isPalindrome(num) {
    const str = String(num);
    return str === [...str].reverse().join('');
}

